orttraining/orttraining/training_ops/cuda/nn/batch_norm_grad.cc

```diff
@@ -5,45 +5,50 @@
 #include "core/providers/common.h"
 #include "core/providers/cuda/cudnn_common.h"
 #include "core/providers/cpu/nn/batch_norm_helper.h"
+#include "core/providers/cuda/math/unary_elementwise_ops_impl.h"
 
 using namespace std;
 namespace onnxruntime {
 namespace cuda {
 
-#define REGISTER_GRADIENT_KERNEL_TYPED(T)                                                  \
-  ONNX_OPERATOR_TYPED_KERNEL_EX(                                                           \
-      BatchNormalizationGrad,                                                              \
-      kMSDomain,                                                                           \
-      1,                                                                                   \
-      T,                                                                                   \
-      kCudaExecutionProvider,                                                              \
-      (*KernelDefBuilder::Create()).TypeConstraint("T", DataTypeImpl::GetTensorType<T>()), \
-      BatchNormalizationGrad<T>);
-
-template <typename T>
-Status BatchNormalizationGrad<T>::ComputeInternal(OpKernelContext* ctx) const {
+#define REGISTER_GRADIENT_KERNEL_TYPED(T, T1, T2)                                            \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                                             \
+      BatchNormalizationGrad,                                                                \
+      kMSDomain,                                                                             \
+      1,                                                                                     \
+      T##_##T1##_##T2,                                                                       \
+      kCudaExecutionProvider,                                                                \
+      (*KernelDefBuilder::Create()).TypeConstraint("T", DataTypeImpl::GetTensorType<T>())    \
+                                   .TypeConstraint("T1", DataTypeImpl::GetTensorType<T1>())  \
+                                   .TypeConstraint("T2", DataTypeImpl::GetTensorType<T2>()), \
+      BatchNormalizationGrad<T, T1, T2>);
+
+template <typename T, typename T1, typename T2>
+Status BatchNormalizationGrad<T, T1, T2>::ComputeInternal(OpKernelContext* ctx) const {
   typedef typename ToCudaType<T>::MappedType CudaT;
+  typedef typename ToCudaType<T1>::MappedType CudaT1;
+  typedef typename ToCudaType<T2>::MappedType CudaT2;
 
   const Tensor* dY = ctx->Input<Tensor>(0);
   const Tensor* X = ctx->Input<Tensor>(1);
   const Tensor* Scale = ctx->Input<Tensor>(2);
   const Tensor* saved_mean = ctx->Input<Tensor>(3);
-  const Tensor* saved_variance = ctx->Input<Tensor>(4);
+  const Tensor* saved_inv_std = ctx->Input<Tensor>(4);
```

Comment:
cudnnBatchnorm kernel actually produce saved_inv_variance, not saved_inv_std.

Reply:
[cudnnBatchNorm](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnBatchNormalizationForwardTraining) claims it is `resultSaveInvVariance`, but in calculation I found that it is actually inv_std of the batch. Should we still change this to be consistent with cudnn doc?

Reply:
Add a manual calculation process below. Data used here are the same as the UT.
<details>
<summary>calculation of BN outputs</summary>

```py
import torch
from torch.nn.parameter import Parameter

x = torch.tensor([-0.2953, 0.1180, 1.0973, -0.1931, -0.1999, -0.0237, 1.5181, 0.0076, -1.0830, -1.5433, 0.4327, -0.9813, 0.7875, -0.4080, -2.3144, 1.5493])
x = x.view(2, 2, 2, 2)
momentum = 0.1
## momentum 0.1 corresponds to momentum 0.9 in PyTorch
m = torch.nn.BatchNorm2d(2, momentum=1 - momentum)
m.running_mean = torch.tensor([1., 2.])
m.running_var = torch.tensor([1., 2.])

use_cuda = True
if use_cuda:
    x = x.cuda()
    m = m.cuda()
    m.weight = Parameter(torch.cuda.FloatTensor([1., 1.]))
    m.bias = Parameter(torch.cuda.FloatTensor([0., 0.]))
    m.running_mean = m.running_mean.cuda()
    m.running_var = m.running_var.cuda()
else:
    m.weight = Parameter(torch.tensor([1., 1.], requires_grad=True))
    m.bias = Parameter(torch.tensor([0., 0.], requires_grad=True))

# manual calculation
batch_mean = torch.mean(x, dim=[0, 2, 3])
print('calculated batch_mean:', batch_mean) # calculated batch_mean: tensor([-0.3060,  0.1146], device='cuda:0')
## same as saved_mean

batch_var_unbiased = ub_var = torch.var(x, dim=[0, 2, 3])
batch_var_biased = torch.var(x, dim=[0, 2, 3], unbiased=False)
print('unbiased batch_var:', batch_var_unbiased) # unbiased batch_var: tensor([0.7569, 1.5405], device='cuda:0')
print('biased batch_var:', batch_var_biased)     # biased batch_var: tensor([0.6623, 1.3480], device='cuda:0')

# inv_var = 1 / var
batch_inv_var_unbiased = torch.reciprocal(batch_var_unbiased)
batch_inv_var_biased = torch.reciprocal(batch_var_biased)
print('unbiased batch_inv_var:', batch_inv_var_unbiased) # unbiased batch_inv_var: tensor([1.3212, 0.6491], device='cuda:0')
print('biased batch_inv_var:', batch_inv_var_biased)     # biased batch_inv_var: tensor([1.5099, 0.7419], device='cuda:0')
## unbiased/biased batch_inv_var is different from saved_inv_std

# inv_std = sqrt(inv_var)
batch_inv_std_unbiased = torch.sqrt(batch_inv_var_unbiased)
batch_inv_std_biased = torch.sqrt(batch_inv_var_biased)
print('unbiased batch_inv_std:', batch_inv_std_unbiased) # unbiased batch_inv_std: tensor([1.1494, 0.8057], device='cuda:0')
print('biased batch_inv_std:', batch_inv_std_biased)     # biased batch_inv_std: tensor([1.2288, 0.8613], device='cuda:0')
## biased batch_inv_std is the same as saved_inv_std

## y calculation
## (x - mean) / sqrt(var) * scale + bias
y_unbiased = (x - batch_mean.view(2, 1, 1)) * batch_inv_std_unbiased.view(2, 1, 1) * m.weight.view(2, 1, 1) + m.bias.view(2, 1, 1)
y_biased = (x - batch_mean.view(2, 1, 1)) * batch_inv_std_biased.view(2, 1, 1) * m.weight.view(2, 1, 1) + m.bias.view(2, 1, 1)
print('y_unbiased:', y_unbiased.view(-1))
print('y_biased:', y_biased.view(-1))
# y_unbiased: tensor([ 0.0123,  0.4874,  1.6130,  0.1298, -0.2534, -0.1114,  1.1308, -0.0862,
#                     -0.8931, -1.4222,  0.8491, -0.7762,  0.5422, -0.4210, -1.9570,  1.1560],
#                     device='cuda:0', grad_fn=<ViewBackward>)
# y_biased: tensor([ 0.0131,  0.5210,  1.7244,  0.1387, -0.2709, -0.1191,  1.2089, -0.0921,
#                   -0.9548, -1.5204,  0.9077, -0.8298,  0.5796, -0.4501, -2.0921,  1.2358],
#                   device='cuda:0', grad_fn=<ViewBackward>)
## When calculating y, the biased batch_inv_std is used.

running_mean = momentum * batch_mean + (1 - momentum) * m.running_mean
running_var_unbiased = momentum * batch_var_unbiased + (1 - momentum) * m.running_var
running_var_biased = momentum * batch_var_biased + (1 - momentum) * m.running_var
print('running_mean:', running_mean)                 # running_mean: tensor([-0.1754,  0.3031], device='cuda:0')
print('running_var_unbiased:', running_var_unbiased) # running_var_unbiased: tensor([0.7812, 1.5865], device='cuda:0')
print('running_var_biased:', running_var_biased)     # running_var_biased: tensor([0.6961, 1.4132], device='cuda:0')
## When calculating running_var, the unbiased batch_var is used.

# PyTorch calculation
y = m(x)

print('y:', y.view(-1)) # y: tensor([ 0.0131,  0.5210,  1.7244,  0.1387, -0.2709, -0.1191,  1.2089, -0.0921,
                        #            -0.9548, -1.5204,  0.9077, -0.8298,  0.5796, -0.4501, -2.0921,  1.2358],
                        #            device='cuda:0', grad_fn=<ViewBackward>)
print('m.running_mean:', m.running_mean) # m.running_mean: tensor([-0.1754,  0.3031], device='cuda:0')
print('m.running_var:', m.running_var)   # m.running_var: tensor([0.7812, 1.5865], device='cuda:0')

```
</details>

Reply:
ah... interesting
if this is the case, then let's keep it as saved_inv_std, and mark it in the comment for this inconsistency.